state/state.go

```diff
@@ -127,12 +127,57 @@ func (s *State) Copy() *State {
 	}
 }
 
+var (
+	errNilGenesisDoc = errors.New("no genesisDoc was found")
+
+	genesisDBKey = []byte("genesis-doc")
+)
+
+// GenesisDoc is the accessor to retrieve the genesisDoc associated
+// with a state. If the state has no set GenesisDoc, it fetches from
+// its database the JSON marshaled bytes keyed by "genesis-doc", and
+// parses the GenesisDoc from that memoizing it for later use.
+// If you'd like to change the value of the GenesisDoc, invoke SetGenesisDoc.
+func (s *State) GenesisDoc() (*types.GenesisDoc, error) {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+
+	if s.cachedGenesisDoc == nil {
+		retrGenesisDocBytes := s.db.Get(genesisDBKey)
+		if len(retrGenesisDocBytes) == 0 {
+			return nil, errNilGenesisDoc
+		}
+		genDoc, err := types.GenesisDocFromJSON(retrGenesisDocBytes)
+		if err != nil {
+			return nil, err
+		}
+		s.cachedGenesisDoc = genDoc
+	}
+	return s.cachedGenesisDoc, nil
+}
+
+func (s *State) ChainID() (string, error) {
+	genDoc, err := s.GenesisDoc()
```

Comment:
same thing. just `return s.chainID`. no errors to return/handle

Reply:
We need to load the genesisDoc from the DB if unset hence the need for it.

Reply:
but we load it the first thing in node.go (https://github.com/tendermint/tendermint/blob/develop/state/state.go#L351), so we can assume that later on, we have it

Reply:
The code you linked is in a function whose comment says that it is used in tests
https://github.com/tendermint/tendermint/blob/925696ca655504bfbc91c314f1a8a357df57960f/state/state.go#L343-L346

Reply:
I initially had it like that but in the description of the feature request @ebuchman says
```
We can provide methods on State to fetch them instead, and we can store them under special keys in the state database
```
which is why I do the fetches in the DB and store them respectively. Perhaps we could discuss the feature more and get a basic spec there, to come to consensus ;)

Reply:
https://github.com/tendermint/tendermint/blob/develop/node/node.go#L141
https://github.com/tendermint/tendermint/blob/develop/state/state.go#L346

the comment should be deleted IMHO

Reply:
Oh, I see. Forget everything I wrote then. I should be more careful about details